[[ Bug 21464-7 ]] Keyboard properties

montegoulding opened the pull request:
This includes #6647 & #6628 due to conflicts so once those are merged this can be rebased.

montegoulding committed:
[[ Bug 21466 ]] Implement returnKeyType field property

This patch adds a new property to fields to control the style of the return
button on soft keyboards on mobile.

runrevmark commented:
@montegoulding Can you redo the commit message / title for the last commit here... Or fold it into one of the others if it just applies to one of them :)

montegoulding commented:
Oh it must have been an stg fail I didn't notice...

montegoulding committed:
[[ Bug 21464 ]] Implement mobileSetKeyboardReturnType

This patch implements `mobileSetKeyboardReturnType` for android and documents
`iphoneSetKeyboardReturnType` as a synonym.

montegoulding committed:
[[ Bug 21467 ]] Implement mobileSetKeyboardDisplay

This patch implements a new command `mobileSetKeyboardDisplay` and function
`mobileGetKeyboardDisplay` to govern the behavior of the keyboard on mobile
to pan the view up if necessary so the focused object is visible.

montegoulding commented:
It turns out I had returnKeyType popped too. It's all fixed up now and changed to 9.5

montegoulding commented:
I have closed the dependant PRs now for simplicity as all commits are here

runrevmark commented:
@montegoulding How does mobileSetReturnKeyType interact with the field property? Is it akin to a global property which the field-level property overrides (if set)?

runrevmark commented on engine/src/field.h:
```diff
@@ -239,7 +240,8 @@ class MCField : public MCControl, public MCMixinObjectHandle<MCField>
     MCTextDirection text_direction;
     MCInterfaceFieldCursorMovement cursor_movement;
-    
+    intenum_t keyboard_type;
```
This should be an explicit MCInterfaceFieldKeyboardType enum it can also be a bit field `: 4` I think should do it.

montegoulding replied:
Hmm... regarding the bitfield I can't help thinking it's risky for an enum because it would be easy to change the enum and not see that the variable is a bitfield. I realise there's a low risk of having more than 16 values in this instance I'm just wondering if we might want to not make this a thing we do...

runrevmark replied:
Hence why I said we needed roundtrip tests for the property values (i.e. repeat for each item tValue in "..."; set the  returnKeyType of field 1 to tValue; TestAssert ..., tValue is the returnKeyType of field 1; end repeat).

montegoulding replied:
aha... woops forgot to add those will do

runrevmark commented on engine/src/field.h:
```diff
@@ -608,6 +610,8 @@ class MCField : public MCControl, public MCMixinObjectHandle<MCField>
     
     MCRectangle firstRectForCharacterRange(int32_t& si, int32_t& ei);
+    
+    intenum_t getkeyboardtype() { return keyboard_type; }
```
This should return MCInterfaceFieldKeyboardType

runrevmark commented on engine/src/field.h:
```diff
@@ -721,6 +725,9 @@ class MCField : public MCControl, public MCMixinObjectHandle<MCField>
     void GetPageHeights(MCExecContext& ctxt, uindex_t& r_count, uinteger_t*& r_heights);
     void GetPageRanges(MCExecContext& ctxt, MCInterfaceFieldRanges& r_ranges);
     
+    void GetKeyboardType(MCExecContext& ctxt, intenum_t& r_type);
```
These remain as intenum_t as they are bound to the property setting logic (which doesn't understand explicit enum types).

runrevmark commented on engine/src/field.h:
```diff
@@ -241,6 +241,7 @@ class MCField : public MCControl, public MCMixinObjectHandle<MCField>
     MCTextDirection text_direction;
     MCInterfaceFieldCursorMovement cursor_movement;
     intenum_t keyboard_type;
+    intenum_t return_key_type;
```
This should be the explicit MCInterfaceReturnKeyType enum - also can be a bitfield.

runrevmark commented:
You might need to make the explicit enums mentioned (which already exist):
```
enum MCInterface... : unsigned
```
By default enums are signed ints which causes annoyance in bitfields.

Additionally can you add tests which check that all the settings roundtrip (i.e. just get/set prop) and unknown ones throw an error.

montegoulding committed:
[[ Bug 21465 ]] Implement keyboardType field property

This patch implements `keyboardType` as a field property. If set the
`keyboardType` will override the keyboard type set with `mobileSetKeyboardType`
for the field.

runrevmark commented:
We also need roundtrip tests for ensuring that each value roundtrips through save/load. e.g.
```
repeat for each item tValue in prop-values
  create field tValue
  set the returnKeyType of it to tValue
end repeat
save stack
delete stack
open stack
repeat for each item tValue in prop-values
  TestAssert the returnKeyType of field tValue is tValue
end repeat
```

montegoulding committed:
[[ Tests ]] Implement reusable enum property value round trip test

This patch adds a reusable enum property value round trip and round trip
to disk test handler to the TestLibrary.

runrevmark commented on engine/src/field.h:
```diff
@@ -239,7 +240,9 @@ class MCField : public MCControl, public MCMixinObjectHandle<MCField>
     MCTextDirection text_direction;
     MCInterfaceFieldCursorMovement cursor_movement;
-    
+    MCInterfaceKeyboardType keyboard_type;
```
These should be bit fields :)

montegoulding replied:
ah.. woops... had some fun with my patches being a bit overlapping and must have forgotten to put those back in. just a sec!

montegoulding replied:
done

runrevmark commented:
@livecode-vulcan review ok dc0a1b3

livecode-vulcan commented:
💙 review by @runrevmark ok dc0a1b383d0c9bdb1ad020557c3425cf2761c83d

livecode-vulcan commented:
😎 test success dc0a1b383d0c9bdb1ad020557c3425cf2761c83d

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/505)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/497)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/497)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/497)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/497)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2019)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/247)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/246)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2051)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2041)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2039)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2074)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2011)
